text/3503-frontmatter.md

```diff
@@ -0,0 +1,512 @@
+- Opens with 3+ backticks and "cargo" followed by a newline
+  - We are intentionally supporting a subset of commonmark code fences in supporting backticks but not tilde code fences or indented fenced code blocks
+  - As we aren't supporting an arbitrarily nested file format (though may show up in one), we likely don't need the flexibility
+  - We are prioritizing on "one right way to do it" to make it easier to learn to write and to read a variety of files.
+- All content is ignored until the same number of backticks or more is found at the start of a line.
+  It is an error to have anything besides spaces and tabs between the backticks and the newline.
+  - Unlike commonmark, it is an error to not close the fenced code block seeing to detect problems earlier in the process seeing as the primary content is what comes after the fenced code block
+
+As cargo will be the first step in the process to parse this,
+the responsibility for high quality error messages will largely fall on cargo.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- A new concept for Rust syntax, adding to overall cognitive load
+- When sharing in markdown documents (e.g. GitHub issues), requires people escape markdown code fences with an extra backtick which they are likely not used to doing (or aware even exists)
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Within this solution,
+we considered starting with only allowing this in the root `mod` (e.g. `main.rs`)
+but decided to allow it in any file mostly for ease of implementation.
+Like with Python, this allows any file in a package (with the correct deps and `mod`s) to be executed, allowing easier interacting experiences in verifying behavior.
+
+As for the hard-coded infostring used by cargo, that is a decision for [RFC 3502](https://github.com/rust-lang/rfcs/pull/3502).
+
+## Syntax
+
+When choosing a syntax for embedding manifests in source files, our care-abouts are
+- How obvious it is for new users when they see it
+- How easy it is for newer users to remember it and type it out
+- How machine editable it is for `cargo add` and friends
+- Needs to be valid Rust code for quality of error messages, etc
+- Leave the door open in case we want to reuse the syntax for embedded lockfiles
+- Leave the door open for single-file `lib`s
+
+This proposed syntax builds off of the precedence of Rust syntax being specialized for an external tool
+(doc-comments for rustdoc).
+However, a new syntax is used instead of extending the comment syntax:
+- Focused on being parsed by arbitrary tools (cargo, vim plugins, etc) without understanding the full Rust grammar
+- Side steps compatibility issues with both
+  user expectations with the looseness of comment syntax (making it harder to parse)
+  and any existing comments that may look like a new structured comment syntax
+
+Benefits for the proposed solution are:
```

Comment:
I'm missing a "Downsides" list for the proposed solution. This feels a bit one-sided compared the the alternatives 😅

Reply:
I can understand.  I normally feel like when there isn't a Downsides, the problem hasn't been thought about enough.  However, this has been an exception so far. 

Frankly, in the initial discussions, none came up.  I just went through all of the pros and cons of the other options, and I don't see any of them applying here as a downside.  I also went back through this thread.  Most comments seem to be more about a preference for another approach, a dislike for this approach, or a philosophical discussions about what justifies syntax support. Maybe I'm not giving enough weight to something but most of the reasons for other approaches or the reasons against come down to details that feel like a stretch to be included in a Downsides.

Reply:
If you're looking for more downsides, let me rephrase parts of my previous comments:

1\. You mention that the familiar syntax is an upside. I consider it a downside, because someone who is familiar with the syntax might assume that it has similar semantics as in markdown. This is not the case:

- The proposed syntax is for specifying metadata, whereas markdown code blocks are part of the document; they just change how the content is displayed (e.g. monospace font, syntax highlighting), so the purpose is completely different
- The proposed syntax is allowed only at the beginning of the file, while markdown code blocks can appear anywhere
- There are several (intentional) syntactic differences between the CommonMark spec and this proposal, which most people won't be aware of

So someone who knows markdown may make incorrect assumptions about this feature. I think another, unfamiliar syntax that doesn't mislead people would be better.

2\. The proposed syntax blurs the line between source code and documentation comments (where markdown is already supported). So people might incorrectly assume that doc-tests can appear outside of doc comments, or that ` ```cargo ` also works inside a doc comment.

3\. Hard-coding the word "cargo" (and TOML as the only supported language) excludes other build tools

4\. Since this syntax is ignored by the compiler, it is effectively a new kind of comment, but Rust already has `//` and `/* */` comments

Reply:
> Hard-coding the word "cargo" (and TOML as the only supported language) excludes other build tools

FWIW, the manifest file in a regular package is also called `Cargo.toml` 😉 

> I just went through all of the pros and cons of the other options, and I don't see any of them applying here as a downside.

as I've mentioned a couple of times now: supporting this new syntax needs significant changes in existing parsers. that seems like a significant downside to me compared to a comment-based approach.

another IMHO big downside is that it introduces another syntax to learn. it might looks roughly like a markdown codefence, but as the previous comment already explained, it is quite restricted and has potential for confusion. doc comments with code fences and those code fences potentially having semantics (doc tests), however is already a concept that Rust users are familiar with. to me it feels like the mental step from doc test comment to cargo manifest comment is smaller than the step from nothing to using code fences directly in the code.

Reply:
> FWIW, the manifest file in a regular package is also called Cargo.toml 

That's fair, but Cargo.toml lives outside of what rustc processes, or the rust language parsing logic. This RFC wants to add `cargo` as a hardcoded string that is whitelisted as allowed, while all other specifiers aren't. That's why I suggested above to disallow non-cargo triple backtick blocks on a *cargo* level: then you don't have to hardcode anything into the compiler, and just error out on a cargo level (which is where this block is being parsed anyways, rust just throws it away as another comment kind). Right now, `cargo` is none of the [compiler builtin symbols](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/symbol/sym/index.html) for example.

Reply:
First off, I overlooked copying the following down into that section

> When sharing in markdown documents (e.g. GitHub issues), requires people escape markdown code fences with an extra backtick which they are likely not used to doing (or aware even exists)

@Aloso 

> So someone who knows markdown may make incorrect assumptions about this feature. I think another, unfamiliar syntax that doesn't mislead people would be better.

I have a slightly different take that I'm phrasing as

> Familiar syntax in an unfamiliar use may make users feel unsettled, unsure how to proceed.

See e0707546

@Aloso 

> 2. The proposed syntax blurs the line between source code and documentation comments (where markdown is already supported). So people might incorrectly assume that doc-tests can appear outside of doc comments, or that ```cargo also works inside a doc comment.

Maybe I'm not seeing this from the right perspective but this feels like a bit of a stretch to me.

@Aloso 

> 3. Hard-coding the word "cargo" (and TOML as the only supported language) excludes other build tools

I think there is a misunderstanding here.  The hard coding is only meant as as starting place as we figure all of this out and is not uncommon in Rust.  For example, last I heard about cfg accessible, they were looking at hard coding it for the standard library.  Rustdoc already has first class syntactic support, with an attribute.  Clippy has a built-in attribute as well.  I've now noted this in 5479104a

> 4. Since this syntax is ignored by the compiler, it is effectively a new kind of comment, but Rust already has // and /* */ comments

Covered this in 7bcaca9e

Reply:
@Turbo87 

> as I've mentioned a couple of times now: supporting this new syntax needs significant changes in existing parsers. that seems like a significant downside to me compared to a comment-based approach.

Yes, and this has been answered multiple times now.  Updates to most parsers is not pressing.  For the parsers that get updated, the change is relatively trivial. 

In addition to those., the Rust grammar is already hard enough to deal with that the number of unique parsers will be relatively low.  That is contrasted with the tools that can operate on `Cargo.toml`.  Frontmatter solutions are the easiest solution for those tools to work with.